### PR TITLE
fix(streaming): resolve word concatenation in streaming output rails

### PR DESCRIPTION
**Pull request by Pouyanpi**
## Problem
The test `test_sequential_streaming_output_rails_allowed` was failing with words being concatenated without spaces:
```
AssertionError: assert 'compliant high quality' in 'This is a safeand complianthigh qualityjoke thatshould passall checks.'
```

## Root Cause
The issue was in `_run_output_rails_in_streaming` method in `llmrails.py`. The method was calling `chunk_str_rep.split()` which removed all whitespace, then tried to reconstruct spacing with `f" {word}"`, but this lost the original token format including trailing spaces.

This created a design contradiction: the buffer strategy creates multi-word chunks (e.g., `"This is a funny "`, `"joke but "`) for efficient output rails processing, but then the code immediately split these back into individual words, defeating the purpose.

## Solution
The fix was to **not split the chunks at all**. The buffer strategy already creates appropriately sized chunks for output rails processing, so we simply yield `user_output_chunks` directly from the buffer strategy.

##  Core Bug Fix
- Remove unnecessary `split()` operations that were breaking word spacing
- Yield chunks directly from buffer strategy preserving original token format
- Fix both `stream_first` and `not stream_first` code paths

##  Interface Improvements  
- **Better API Design**: Refactored `BufferStrategy` with `process_stream()` method and `ChunkBatch` named tuple
- **Clear Separation**: Distinguished between `processing_context` (for output rails) and `user_output_chunks` (for streaming)
- **Named Tuple**: Replaced opaque `Tuple[List[str], List[str]]` with self documenting `ChunkBatch`

## Test Coverage Improvements
- Enhanced `tests/test_buffer_strategy.py` from 2 basic tests to 11 comprehensive tests
- Added edge case coverage (empty streams, boundary conditions, input validation)
- Added realistic token data with proper spacing validation
- Added interface testing (`process_stream()` vs `__call__()` parity)

### CODE REVIEW

File: nemoguardrails/rails/llm/buffer.py [hunk 1]
```diff
@@ -14,95 +14,379 @@
...
+            >>> result = strategy.format_chunks(chunks)
+            >>> print(result)  # "Hello world"
+        """
+        ...
 
     @abstractmethod
-    def generate_chunk_str(self, *args, **kwargs) -> str:
-        pass
+    async def process_stream(
+        self, streaming_handler
```
Comment (trebedea) [inline nemoguardrails/rails/llm/buffer.py:115]:
```suggestion
        self, streaming_handler: AsyncIterator[ChunkBatch]
```

File: nemoguardrails/rails/llm/buffer.py [hunk 2]
```diff
@@ -14,95 +14,379 @@
...
+            ...     # for output rails processing (needs context):
+            ...     context_formatted = strategy.format_chunks(chunk_batch.processing_context)
+            ...     # for user output (new content only):
+            ...     user_formatted = strategy.format_chunks(chunk_batch.user_output_chunks)
+            ...     print(f"Processing: {context_formatted}")
+            ...     print(f"User: {user_formatted}")
+        """
+        ...
+
+    async def __call__(self, streaming_handler) -> AsyncGenerator[ChunkBatch, None]:
```
Comment (trebedea) [inline nemoguardrails/rails/llm/buffer.py:143]:
```suggestion
    async def __call__(self, streaming_handler: AsyncIterator[ChunkBatch]) -> AsyncGenerator[ChunkBatch, None]:
```

File: nemoguardrails/rails/llm/buffer.py [hunk 3]
```diff
@@ -14,95 +14,379 @@
...
-        # Yield any remaining buffer if it's not empty
+        # yield any remaining buffer if it's not empty
         if buffer:
-            yield (
-                buffer,
-                self.generate_chunk_str(
-                    buffer[-self.buffer_chunk_size - self.buffer_context_size :], index
-                ),
+            new_chunks_to_yield = len(buffer) - max(
+                0, self.total_yielded - (total_chunks - len(buffer))
```
Comment (trebedea) [inline nemoguardrails/rails/llm/buffer.py]:
```suggestion
            new_chunks_to_yield = min(len(buffer), total_chunks - self.total_yielded)
```

File: nemoguardrails/rails/llm/buffer.py [hunk 4]
```diff
@@ -14,95 +14,379 @@
...
             )
+            if new_chunks_to_yield > 0:
+                chunks_to_yield = (
+                    buffer[-new_chunks_to_yield:]
+                    if new_chunks_to_yield <= len(buffer)
```
Comment (trebedea) [inline nemoguardrails/rails/llm/buffer.py]: Seems like this is always true as 

`new_chunks_to_yield = min(len(buffer), total_chunks - self.total_yielded)` from above

File: nemoguardrails/rails/llm/llmrails.py [hunk 5]
```diff
@@ -1056,17 +1056,39 @@ def stream_async(
...
+                )
+            except Exception as e:
+                # If an exception occurs during generation, push it to the streaming handler
+                # This ensures the streaming pipeline is properly terminated
+                log.error(f"Error in generation task: {e}", exc_info=True)
+                error_message = str(e)
+                error_dict = extract_error_json(error_message)
+                error_payload = json.dumps(error_dict)
+                await streaming_handler.push_chunk(error_payload)
+                await streaming_handler.push_chunk(END_OF_STREAM)
```
Comment (trebedea) [inline nemoguardrails/rails/llm/llmrails.py:1077]: Nice!

File: nemoguardrails/rails/llm/llmrails.py [hunk 6]
```diff
@@ -1056,17 +1056,39 @@ def stream_async(
...
+            try:
+                await self.generate_async(
+                    prompt=prompt,
+                    messages=messages,
+                    streaming_handler=streaming_handler,
+                    options=options,
+                    state=state,
+                )
+            except Exception as e:
+                # If an exception occurs during generation, push it to the streaming handler
```
Comment (trebedea) [inline nemoguardrails/rails/llm/llmrails.py]:
```suggestion
                # If an exception occurs during generation, push it to the streaming handler as a json string 
```

File: nemoguardrails/rails/llm/llmrails.py [hunk 7]
```diff
@@ -1377,45 +1399,46 @@ def _prepare_params(
...
-            chunk_str = " ".join(chunk_list)
+        async for chunk_batch in buffer_strategy(streaming_handler):
+            user_output_chunks = chunk_batch.user_output_chunks
+            # format processing_context for output rails processing (needs full context)
+            bot_response_chunk = buffer_strategy.format_chunks(
+                chunk_batch.processing_context
+            )
+
+            # check if user_output_chunks is a list of individual chunks
+            # or if it's a JSON error string
```
Comment (trebedea) [inline nemoguardrails/rails/llm/llmrails.py]:
```suggestion
            # or if it's a JSON string, by convention this means an error occurred and the error dict is stored as a JSON
```

**Review by trebedea — APPROVED** [review]
Looks good, just some minor comments! 
Great to have this fix. 👍

**Comment by Pouyanpi** [comment]
Thank you @trebedea for the review. I applied your suggestions, it made it more readable 👍🏻

File: tests/test_buffer_strategy.py [hunk 8]
```diff
@@ -24,12 +26,40 @@ async def fake_streaming_handler():
...
+    yield  # unreachable
+
+
 @pytest.mark.asyncio
 async def test_buffer_strategy():
     buffer_strategy = BufferStrategy(buffer_context_size=5, buffer_chunk_size=10)
     streaming_handler = fake_streaming_handler()
 
-    expected_buffers = [
+    expected_processing_contexts = [
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply tests/test_buffer_strategy.py:65]: outdated:

line 57: Uses RollingBuffer correctly (not BufferStrategy)
lines 59-79: Properly defines 3 expected processing contexts
lines 81-97: Properly defines 3 expected user output chunks

**Comment by Pouyanpi** [comment]
> ## [Codecov](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA) Report
> Attention: Patch coverage is `81.25000%` with `12 lines` in your changes missing coverage. Please review.
> 
> > Project coverage is 69.64%. Comparing base [(`9cdad05`)](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/commit/9cdad053b670ff88e7a8fa67f950af36dab5496a?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA) to head [(`58edd2f`)](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/commit/58edd2ff9aa79e36213e4d9c4bd3de33ecb5c766?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA).
> > Report is 7 commits behind head on develop.
> 
> [Files with missing lines](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA)	Patch %	Lines
> [nemoguardrails/rails/llm/llmrails.py](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?src=pr&el=tree&filepath=nemoguardrails%2Frails%2Fllm%2Fllmrails.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA#diff-bmVtb2d1YXJkcmFpbHMvcmFpbHMvbGxtL2xsbXJhaWxzLnB5)	71.87%	[9 Missing ⚠️ ](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA)
> [nemoguardrails/rails/llm/buffer.py](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?src=pr&el=tree&filepath=nemoguardrails%2Frails%2Fllm%2Fbuffer.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA#diff-bmVtb2d1YXJkcmFpbHMvcmFpbHMvbGxtL2J1ZmZlci5weQ==)	90.62%	[3 Missing ⚠️ ](https://app.codecov.io/gh/NVIDIA/NeMo-Guardrails/pull/1259?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=NVIDIA)
> Additional details and impacted files
> 🚀 New features to boost your workflow:

it is using outdated code for llmrails.py

it does not understand that our tests cover 4 missing lines from buffer.py
